### ISSUE TITLE
2 gang Switch module (_TZ3000_zmy4lslw - TS0002)

### ISSUE THREAD

**Issue by Paalap** [dresden-elektronik/deconz-rest-plugin#6446]
## Device

- Product name: 2CH Zigbee Switch Module (Tuya Smart ZigBee 3.0 Switch Module)
- [AliExpress](https://de.aliexpress.com/item/1005004371232755.html)
- Manufacturer: _TZ3000_zmy4lslw 
- Model identifier: TS0002
- Device type : 2 gang/ch Switch

The device is connected reports the states properly but switching on/off is always applied to both channels (fun fact  the 4ch model from the same source above (with _TZ3000_ltt60asa / TS0004) is working fine).

## Screenshots

![image](https://user-images.githubusercontent.com/42343574/197199997-4273e386-06e3-4337-8cae-d0c4964bba4d.png)

### Basic
![image](https://user-images.githubusercontent.com/42343574/197200177-8378ecf8-fe96-400d-8077-c96f13478f06.png)

### Identify
![image](https://user-images.githubusercontent.com/42343574/197200233-ecb35b7f-aab4-4490-82cb-1b400561954b.png)

### Groups
![image](https://user-images.githubusercontent.com/42343574/197200268-120b0837-2f55-4190-bc11-f2dbd659323c.png)

### Scenes
![image](https://user-images.githubusercontent.com/42343574/197200308-952eea1f-2484-43b8-938c-dfb2948fae07.png)
![image](https://user-images.githubusercontent.com/42343574/197200375-091b3265-f4ea-4051-bfdd-8c7b582e5cc3.png)
![image](https://user-images.githubusercontent.com/42343574/197200406-530ca5bc-af37-477e-87a4-ad17b7336393.png)

### On/Off
1st
![image](https://user-images.githubusercontent.com/42343574/197200432-3cc0796c-a92a-4d3b-84c8-c1da145d1207.png)

2nd
![image](https://user-images.githubusercontent.com/42343574/197200534-3655f618-e062-41cc-b1e0-88cbbb1a2737.png)

### Simple Metering
![image](https://user-images.githubusercontent.com/42343574/197200573-ce1b1555-fb34-4f17-a183-99118bf427cc.png)

Thank you in advance!

**Comment by Paalap** [comment]
Is there a way to support the development?

There is/was a [similar issue on zigbee2mgtt](https://github.com/Koenkk/zigbee2mqtt/issues/14523#issue-1412091591) - just in case this is helpful

And would it it be possible to expose "power on behavior"?

Thanks

**Comment by whit33r** [comment]
I have a 3 CH version of this switch, and that have same issue.

<img width="810" alt="Képernyőfotó 2022-10-25 - 15 42 29" src="https://user-images.githubusercontent.com/866757/197789589-ae9e4750-41ce-4922-a290-af791fd7d6fa.png">

<img width="354" alt="Képernyőfotó 2022-10-25 - 13 50 56" src="https://user-images.githubusercontent.com/866757/197789247-05651541-8bd1-453b-9892-cdd47b2e53bc.png">

<img width="263" alt="Képernyőfotó 2022-10-25 - 15 44 30" src="https://user-images.githubusercontent.com/866757/197790155-77632e2d-6871-4119-88c6-d44856c09073.png">

**Comment by Paalap** [comment]
Would be happy to keep it active

**Comment by Smanar** [comment]
This device need a "tuya unlock sequence" to work.
So not possible to add it for the moment.

**Comment by Smanar** [comment]
Nope, need to add the feature on c++ side.
And as its not a priority ...

**Comment by OndraM** [comment]
@Smanar I don't see `0xfffe` attribute in the deconz GUI in the basic cluster, is there any manual way how to show it?

Or maybe is it possible to send the write commands via some API? This way it would be possible to send the sequence.

**Comment by Smanar** [comment]
>Or maybe is it possible to send the read commands via API?

Nope, sorry, the only way is using the GUI or the c++ code.

>is there any manual way how to show it?

If you are not using HA yes (I don't think this file was persistent, but have never tried), there is an xml file in your config folder called general.xml

I dont see the manufacture number on the capture, it's visible on the "node information" but probably can miss here

```
		<attribute-set id="0xFF00" description="Xiaomi specific" mfcode="0x115f">
			<attribute id="0xFF08" name="Unknown" type="u16" access="rw" required="o" mfcode="0x115f"></attribute>
			<attribute id="0xFF51" name="Unknown" type="ostring" access="rw" required="o" mfcode="0x115f"></attribute>
			<attribute id="0xFF52" name="Unknown" type="cstring" access="rw" required="o" mfcode="0x115f"></attribute>
			<attribute id="0xFF53" name="Unknown" type="ostring" access="rw" required="o" mfcode="0x115f"></attribute>
			<attribute id="0xFF54" name="Unknown" type="ostring" access="rw" required="o" mfcode="0x115f"></attribute>
			<attribute id="0xFFF0" name="Unknown" type="ostring" access="rw" required="o" mfcode="0x115f"></attribute>
		</attribute-set>
```

**Comment by ranelazari** [comment]
Hi,
is there ny devalopment of thias issue?

**Comment by OndraM** [comment]
@Smanar Ok, thanks. Any idea how to know whether the sequence succeeded? How can I recognize the device started to work normally after receiving the unlock sequence? Will this "just happen" or I need for example to switch some attribute to enable the "individual switch" mode?

**Comment by Smanar** [comment]
Ha yes right.
No it's realy something magic, it work or not. return are classic.

You can use the GUI, choose a endpoint "01" or "02", select the cluster 0x0006 on/off, and in the cluster panel try the "on" or "off" command.
If it have action on 1 channel/part/gang it have worked, else it will switch the 2 channel/part/gang

**Comment by OndraM** [comment]
Ok, so I spend like two hours on this, and I'm pretty confident it cannot be done using UI. 

- I edited general.xml, added the `0xfffe` attribute to the basic cluster. I removed all other attributes in the cluster, leaving just those required for tuya magic spell, in the right order.
- Then I copied the file into deconz docker image on startup, to override the default one.
- I opened the Basic cluster of the device in the UI.
- I tried using the "Read" button above all attributes (because it then reads all attributes in the cluster) but with no luck. The logs shows that deconz is sending request to some other attributes right after it goes through the visible, so maybe it breakes the signature? Who knows.
- I tried going through those six attributes and pressing read individually in their respective windows, with no luck.
- I even write script ("macro") which does the same automatically. Even though it is able to cycle through all the attributes in like ~two seconds, there is still no change in behavior of the device.

So either I'm missing something (I don't know when should this sequence be played? Right when the device is pairing? Or anytime later?). Or the two seconds are just to slow and it don't get recognized by the device and don't unlock it.


Also, I really thing this should be prioritized in the deconz, as there is already ~80 devices (according to z2m source code) by various manufacturers (tuya but also lidl, moes, zemismart...) using the magic packet, which are broken because of this.

**Comment by OndraM** [comment]
Also, I found [this reference](https://community.smartthings.com/t/four-button-tuya-zigbee-device-dth-dth-link-in-post-7-same-device-from-multiple-brands-zemismart-yagusmart-losonho-dyhf-etc/206149/120):

> It is important to send the readAttribute commands for the Basic and Power clusters at once. Sending individual readAttribute commands for each attribute didn’t work.

So yes, it seems its not possible to achieve this in deconz UI, because it sends the commands individually.

**Comment by ThiemeNL** [comment]
I have the same issue with [_TZ3000_bvrlqyj7](https://forum.phoscon.de/t/tuya-tz3000-bvrlqyj7-switch/2427) sending an ON command is turning on both channels.

**Comment by GitMouth** [comment]
Hi, any suggestion to make them works ? ....
_TZ3000_yf8iuzil  TS0002
Same behavior....

**Comment by Linqman** [comment]
I have the same issue with TS0012 _TZ3000_jl7qyupf, Switch Module 2 Channel No Neutral

**Comment by OndraM** [comment]
> Hi, any suggestion to make them works ? ....

I solved it by migrating to zigbe2mqtt. According to information Smanar gave (that it won't be worked on this anytime soon) this is the only solution. I spend many hours trying to get this work with deConz, but finally I gave up and switched to zigbe2mqtt. So far, I didn't have a single issue, everything works fine and I definitely won't regret the extra time the migration took.

**Comment by Paalap** [comment]
Same here, running 2 ZigBee networks in parallel. Not what I really like but it is ok for now.

**Comment by micsey** [comment]
I have the same issue with 4 gang Switch module ( _TZ3000_ltt60asa - TS0004). In zigbee2mqtt i can find this device as supported.

**Comment by GitMouth** [comment]
thx for your experience. I'm working with Deconz, and don't want to reappair and make my automations with mqtt... I'll selling them...

**Comment by ThiemeNL** [comment]
Please keep this issue open

**Comment by BabaIsYou** [comment]
Hi folks, 

@Smanar did a great job in doing this Tuya Magic sequence to be integrated in a [DDF](https://forum.phoscon.de/t/including-standard-tuya-magic-spell-sequence-during-joining/2916/15). That's perhaps juste a workaround waiting for something more optimized but can try to add the partin the referenced issue in DDF.

**Comment by Smanar** [comment]
Probably something like

```
{
  "schema": "devcap1.schema.json",
  "manufacturername": "_TZ3000_zmy4lslw",
  "modelid": "TS0002",
  "product": "Tuya 2 output module",
  "sleeper": false,
  "status": "Gold",
  "subdevices": [
    {
      "type": "$TYPE_ON_OFF_OUTPUT",
      "restapi": "/lights",
      "uuid": [
        "$address.ext",
        "0x01"
      ],
      "items": [
        {
          "name": "attr/id"
        },
        {
          "name": "attr/lastannounced"
        },
        {
          "name": "attr/lastseen"
        },
        {
          "name": "attr/manufacturername"
        },
        {
          "name": "attr/modelid"
        },
        {
          "name": "attr/name"
        },
        {
          "name": "attr/swversion",                   
          "refresh.interval": 86400,
          "parse": {"fn": "zcl", "ep": 1, "cl": "0x0000", "at": "0x0001", "script": "tuya_swversion.js"},
          "read": {"fn": "zcl", "ep": 1, "cl": "0x0000", "at": "0x0001"}
        },
        {
          "name": "attr/type"
        },
        {
          "name": "attr/uniqueid"
        },
        {
          "name": "state/on"
        },
        {
          "name": "state/reachable"
        },
        {
          "name": "config/interfacemode",
          "refresh.interval": 86400,
          "read": {
            "fn": "zcl",
            "ep": 1,
            "cl": "0x0000",
            "at": [
              "0x0004",
              "0x0000",
              "0x0001",
              "0x0005",
              "0x0007",
              "0xfffe"
            ]
          },
          "parse": {
            "fn": "zcl",
            "ep": 1,
            "cl": "0x0000",
            "at": "0xfffe",
            "eval": "Item.val = Attr.val"
          },
          "public": false,
          "default": 0
        }
      ]
    },
    {
      "type": "$TYPE_ON_OFF_OUTPUT",
      "restapi": "/lights",
      "uuid": [
        "$address.ext",
        "0x02"
      ],
      "items": [
        {
          "name": "attr/id"
        },
        {
          "name": "attr/lastannounced"
        },
        {
          "name": "attr/lastseen"
        },
        {
          "name": "attr/manufacturername"
        },
        {
          "name": "attr/modelid"
        },
        {
          "name": "attr/name"
        },
        {
          "name": "attr/swversion",                   
          "refresh.interval": 86400,
          "parse": {"fn": "zcl", "ep": 1, "cl": "0x0000", "at": "0x0001", "script": "tuya_swversion.js"},
          "read": {"fn": "zcl", "ep": 1, "cl": "0x0000", "at": "0x0001"}
        },
        {
          "name": "attr/type"
        },
        {
          "name": "attr/uniqueid"
        },
        {
          "name": "state/on"
        },
        {
          "name": "state/reachable"
        }
      ]
    }
  ]
}
```

**Comment by BlauerPulli** [comment]
hey, just leave a comment to re-open this again. have the same device and it also not working for me. i will test @Smanar s suggestion soon.

**Comment by ThiemeNL** [comment]
I tested @Smanar suggestion, but it's not working for me.

**Comment by Smanar** [comment]
Uou have re-included the device ? (else need to wait 24h)

**Comment by commonslibs** [comment]
Same problem here.

**Comment by Smanar** [comment]
You have used the new DDF and even after a re-inclusion both channels are linked ?

**Comment by ThiemeNL** [comment]
Yes, I tried the new DDF and re-inclusion. The behavior is now different. Looks like one channel is still connected to both outputs and but the other one channel is not working anymore. Reading the channel status is working fine.

**Comment by commonslibs** [comment]
How can I put that DDF on my system? Any links that explain it? 

By the way, I understand that the "issue" should have an "open" state and not like the one that has to be "closed", right?

**Comment by ThiemeNL** [comment]
Maybe this link is helpful: https://dresden-elektronik.github.io/deconz-dev-doc/modules/ddf/

For me its a while ago. I've upload the DDF file to `/home/{user}/.local/share/dresden-elektronik/deCONZ/devices`

The ticket was automatically closed due to inactivity.

**Comment by commonslibs** [comment]
Thank you all very much and especially @Smanar and @ThiemeNL

The problem is already **solved**.

I have uploaded the DDF file that @Smanar had posted some answers before and it has worked perfectly for me.

Regarding the file that @Smanar put, I have modified very few things in the initial part of it, especially to put the vendor and some other minor change, which I understand has not been what has solved the problem, but I liked leaving the DDF more clear.

Once I have created the DDF file with the extension ".json" I have uploaded it to the folder "..."\devices\WETEN\TZ3000_zmy4lslw.json" and I have restarted the deconz docker container.

Once the container has restarted I have gone to home assistant and now I have been able to turn the bulbs on and off individually without problems.

I paste the screenshot of the DDF in deconz:
![image](https://user-images.githubusercontent.com/43066090/236613271-f0f97bce-8a1a-4341-85b7-d86010ce02b4.png)

I attach the DDF file with the small changes that I have made.

[TZ3000_zmy4lslw.zip](https://github.com/dresden-elektronik/deconz-rest-plugin/files/11411900/TZ3000_zmy4lslw.zip)

As said, thank you very much.

**Comment by Smanar** [comment]
>By the way, is it possible that in future versions of "deconz" this device will be integrated without problems without having to carry out this "workaround" that consists of uploading the DDF file to "deconz"?

Yep, ofc https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6953

But I haven't used your modification because tuya is just clone of clones of clones, so better to use generic DDF.
Have used too a simplified config for the tuya unbclock command, so you can have an unused field now on your side ("config/tuya_unlock" instead of "config/interfacemode")

**Comment by ThiemeNL** [comment]
Unfortunately your solution is not working on the`_TZ3000_bvrlqyj7` 
The device has a `Tuya special light mode` which is set to `Mode 1` ![image](https://user-images.githubusercontent.com/5397828/236675333-e87cce12-d2c8-43e7-b6d7-baef1ba67601.png) 
I think this is preventing the solution to work, since the 2 channel mode is set on the devices it selves. 
Any advise how to change the light mode?

**Comment by BabaIsYou** [comment]
Did you try to change it into deconz GUI ?

**Comment by ThiemeNL** [comment]
Yes, I tried, but it's a read only field.

**Comment by Smanar** [comment]
Nothing special for your devices, all of them are clone and use the same code on Z2M

```
        fingerprint: [{modelID: 'TS0002', manufacturerName: '_TZ3000_01gpyda5'}, {modelID: 'TS0002', manufacturerName: '_TZ3000_bvrlqyj7'},
            {modelID: 'TS0002', manufacturerName: '_TZ3000_7ed9cqgi'}, {modelID: 'TS0002', manufacturerName: '_TZ3000_zmy4lslw'}],
```

Can you share the used DDF ?

**Comment by ThiemeNL** [comment]
Currently I have loaded this file: https://gist.githubusercontent.com/ThiemeNL/ce8f0db07da2c4d78b99ad7ab576ccfe/raw/990585a12a6ed5fe6519a639db18df2656287a3f/TZ3000_bvrlqyj7.json
I change the vendor to something custom, so I could see that my changes where loaded.

I also tried your PR version and only changed the `manufacturername` to `_TZ3000_bvrlqyj7` but it didn't help
https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6953/commits/8abab56069e65076cbdd157e65f4e8ddb3d7b824#diff-3051f341cfb05084eed8f8c89b841ebc3ed4430fab03b4da6047cd1fc781f650

**Comment by Smanar** [comment]
The model id is "TS0002" too ?

**Comment by ThiemeNL** [comment]
Yes, I deleted the lights, rebooted my Pi and then added the lights again: 
![image](https://user-images.githubusercontent.com/5397828/236699006-c914aade-2248-4d73-80ff-156081e2138a.png)

**Comment by Mimiix** [comment]
Please open a seperate request for this, as this is not the same device.

**Comment by sciorty** [comment]
I have the same problem with this device _TZ3000_nuenzetq:

<img width="358" alt="Schermata 2024-02-17 alle 20 41 00" src="https://github.com/dresden-elektronik/deconz-rest-plugin/assets/6460914/7175f276-96ad-4a3c-a828-9da07eeee84f">

<img width="715" alt="Schermata 2024-02-17 alle 20 41 53" src="https://github.com/dresden-elektronik/deconz-rest-plugin/assets/6460914/00ebe574-979a-4086-84a3-7a63aba09b38">

**Comment by Smanar** [comment]
You mean not reconised ?
You can try with adding the modelID+Manufacture name on the DDF.

**Comment by sciorty** [comment]
> You mean not reconised ? You can try with adding the modelID+Manufacture name on the DDF.

I mean that when i try to turn one of the switches on, they both turn on.
If the solution would still be to add it into the ddf, which ddf is it?

**Comment by sciorty** [comment]
It’s Ts0002 _TZ3000_nuenzetq with 2 gangs.
So i just have to add this json file where? Even if it’s named differently? Or do i have to change/add something on the inside?

**Comment by Smanar** [comment]
As you want, depend of OS.
You can just add your model

```
  "manufacturername": ["_TZ3000_zmy4lslw", "_TZ3000_nuenzetq"],
  "modelid": ["TS0002", "TS0002"],
```

On the actual DDF.
Else  (if you have docker for exemple) you need to create a new DDF in a different folder.

> All DDF provided directly with deCONZ typically reside in /usr/share/deCONZ/devices/ on a Linux system and are loaded first. However, files residing in the home directory of the user running deCONZ (e.g. /home/<DECONZUSER>/.local/share/dresden-elektronik/deCONZ/devices) will override the pre-packaged files to allow users to amend and keep their own files if desired.

**Comment by sciorty** [comment]
Sorry that's my lack but I miss some informations. This is what I see if I try to edit the DDF through deCONZ
<img width="826" alt="Schermata 2024-02-19 alle 22 58 21" src="https://github.com/dresden-elektronik/deconz-rest-plugin/assets/6460914/fb8d73e4-51c8-41cd-9300-b0d34af6cfb2">
<img width="183" alt="Schermata 2024-02-19 alle 22 58 06" src="https://github.com/dresden-elektronik/deconz-rest-plugin/assets/6460914/015aea3c-e5bb-4335-82f6-0e6f30222334">


Should I modify it? Or should I create a file here (Home assistant supervised)?
<img width="779" alt="Schermata 2024-02-19 alle 23 00 49" src="https://github.com/dresden-elektronik/deconz-rest-plugin/assets/6460914/bf77d14d-c6b1-4d2d-a42b-5b08e1e25e10">
How the file should be created? Just a copy/paste of [this](https://github.com/Smanar/deconz-rest-plugin/blob/581120bc632b4b35068f3edba37c8b08b4d7623f/devices/tuya/_TZ3000_zmy4lslw_2channel_module.json) ?

**Comment by sciorty** [comment]
OK now it's working. I created the ddf on the path that I mentioned and called it _TZ3000_nuenzetq_2channel_module.json
Took [this ddf](https://github.com/Smanar/deconz-rest-plugin/blob/581120bc632b4b35068f3edba37c8b08b4d7623f/devices/tuya/_TZ3000_zmy4lslw_2channel_module.json) that apparently is the same device but of a different vendor (right?) and just changed the "manufacturername" field with mine.
On deCONZ I went to the same screenshot I posted and gone to File->Open and selected this newly created file so that it is loaded as a ddf for the device. Then did a Hot reload and it's working!

Is this the processo on how to manually "set" the ddf of the devices when I have some problems? How to simply modify the existing ones in case I have problems? What's the difference in the file's folder locations? How do I know that this device will be using the ddf I loaded? Eg: future releases will contain this .json and my "manually" created will not be used anymore and I can delete it?

**Comment by Smanar** [comment]
>Is this the processo on how to manually "set" the ddf of the devices when I have some problems?

Depend of OS, HA supervised don't use docker ?
On normal OS deconz will use 2 "devices" folder, 1 is for user, untouched (so you will need to remove the file yourself), the second one is reset a every deconz update, on docker it's another story.
Make a right clic, then "edit DDF", on the editor titlebar you will see the DDf file used.

Have added your device in list  https://github.com/dresden-elektronik/deconz-rest-plugin/issues/6446
